regnety/models/blocks.py

```diff
@@ -0,0 +1,219 @@
+            momentum = 0.9, epsilon = 0.00001)
+
+        self.relu1x1_1 = tf.keras.layers.ReLU()
+        self.relu3x3 = tf.keras.layers.ReLU()
+        self.relu1x1_2 = tf.keras.layers.ReLU()
+
+        self.skip_conv = None
+        self.conv3x3 = None
+        self.bn_skip = None
+        self.relu_skip = None
+
+        if (in_filters != out_filters) or (stride != 1):
+            self.skip_conv = tf.keras.layers.Conv2D(
+                out_filters, (1, 1), strides=2)
+            self.bn_skip = tf.keras.layers.BatchNormalization()
+            self.relu_skip = tf.keras.layers.ReLU()    
+ 
+            self.conv3x3 = tf.keras.layers.Conv2D(
+                out_filters, (3, 3), strides = 2, groups = self.groups, 
+                padding = 'same')
+ 
+        else:
+            self.conv3x3 = tf.keras.layers.Conv2D(
+                out_filters, (3, 3), strides = 1, groups = self.groups, 
+                padding = 'same')
+        
+    
+    def call(self, inputs):
+        x = self.conv1x1_1(inputs)
+        x = self.bn1x1_1(x)
+        x = self.relu1x1_1(x)
+
+        x = self.conv3x3(x)
+        x = self.bn3x3(x)
+        x = self.relu3x3(x)
+        
+        x = self.se(x)
+        
+        x = self.conv1x1_2(x)
+        x = self.bn1x1_2(x)
+        x = self.relu1x1_2(x)
+
+        if self.skip_conv is not None:
+            skip_tensor = self.skip_conv(inputs)
+            skip_tensor = self.bn_skip(skip_tensor)
+            skip_tensor = self.relu_skip(skip_tensor)
+        
+        else:
+            skip_tensor = inputs
+        
+        x = x + skip_tensor
+
+        return x
+    
+
+
+class Stage(tf.keras.layers.Layer):
+    """
+    Class for RegNetY stage. 
+
+    Args:
+        depth: Depth of stage, number of blocks to use
+        group_width: Group width of all blocks in  this stage
+        in_filters: Input filters to this stage
+        out_filters: Output filters from this stage
+        
+    """
+
+    def __init__(self,
+        depth:int,
+        group_width:int,
+        in_filters:int,
+        out_filters:int
+    ):
+        super(Stage, self).__init__()
+        
+        self.depth = depth
+
+        self.stage = []
+
+        self.stage.append(YBlock(group_width, in_filters, out_filters, stride = 2))
+
+        for _ in range(depth - 1):
+            self.stage.append(
+                YBlock(group_width, out_filters, out_filters, stride = 1)
+            )
+
+    def call(self, inputs):
+        x = inputs
+        for i in range(self.depth):
+            x = self.stage[i](x)
+        
+        return x
+
+
+class Head(tf.keras.layers.Layer):
+    """
+    Head for all RegNetY models.
+
+    Args:
+        num_classes: Integer specifying number of classes of data. 
+    """
+    def __init__(self, num_classes):
+        super(Head, self).__init__()
+
+        self.gap = tf.keras.layers.GlobalAveragePooling2D()
+        self.fc = tf.keras.layers.Dense(num_classes, activation = 'sigmoid')
```

Comment:
`sigmoid` for a multi-class classification problem? Shouldn't this be softmax? Also, let's separate out the last layer activation a distinct layer to give the users a chance to operate on the raw logits.

Reply:
Sure, that's what they did in the official implementation.

Reply:
Yeah but the `sigmoid` part is definitely not correct unless I am missing out on something.

Reply:
> Or we can omit the softmax layer totally and specify from_logits=True in Categorical_Crossentropy().

I was replying to this. Sigmoid is incorrect, I agree.